File: practice/weaks/2-3/DAST/authirization/2fa_test.go
package main

import (
    "bytes"
    "encoding/json"
    "fmt"
    "io/ioutil"
    "net/http"
    "time"

    "github.com/pquerna/otp/totp"
)

// Shared secret из QR-кода
const secret = "JNJAM7A7PRSGMFRN"

// URL для логина и 2FA
const authURL = "http://localhost:9092/rest/user/login"
const twoFaURL = "http://localhost:9092/rest/2fa/verify"

// Логин и пароль
const username = "test@mail.com"
const password = "password123"

// Генерация 2FA-кода с использованием pquerna/otp
func generate2FACode(secret string) string {
    twoFaCode, err := totp.GenerateCode(secret, time.Now())
    if err != nil {
        fmt.Printf("❌ Ошибка генерации 2FA-кода: %v\n", err)
        return ""
    }
    fmt.Printf("✅ Сгенерированный 2FA-код: %s\n", twoFaCode)
    return twoFaCode
}

// Выполнение первого этапа авторизации (логин)
func login(authURL, username, password string) map[string]interface{} {
    payload := map[string]string{
        "email":    username,
        "password": password,
    }
    payloadBytes, _ := json.Marshal(payload)

    resp, err := http.Post(authURL, "application/json", bytes.NewBuffer(payloadBytes))
    if err != nil {
        fmt.Printf("❌ Ошибка логина: %v\n", err)
        return nil
    }
    defer resp.Body.Close()

    body, _ := ioutil.ReadAll(resp.Body)
    if resp.StatusCode != http.StatusOK {
        fmt.Printf("❌ Ошибка логина: %d\n%s\n", resp.StatusCode, string(body))
        return nil
    }

    fmt.Println("✅ Успешный логин")
    var result map[string]interface{}
    json.Unmarshal(body, &result)
    return result
}

// Выполнение второго этапа авторизации (2FA)
func send2FA(twoFaURL, twoFaCode string) map[string]interface{} {
    payload := map[string]string{
        "code": twoFaCode,
    }
    payloadBytes, _ := json.Marshal(payload)

    resp, err := http.Post(twoFaURL, "application/json", bytes.NewBuffer(payloadBytes))
    if err != nil {
        fmt.Printf("❌ Ошибка 2FA: %v\n", err)
        return nil
    }
    defer resp.Body.Close()

    body, _ := ioutil.ReadAll(resp.Body)
    if resp.StatusCode != http.StatusOK {
        fmt.Printf("❌ Ошибка 2FA: %d\n%s\n", resp.StatusCode, string(body))
        return nil
    }

    fmt.Println("✅ Успешная 2FA")
    var result map[string]interface{}
    json.Unmarshal(body, &result)
    return result
}

func main() {
    // Первый этап: логин
    loginResponse := login(authURL, username, password)
    if loginResponse == nil {
        return
    }

    // Генерация 2FA-кода
    twoFaCode := generate2FACode(secret)
    if twoFaCode == "" {
        return
    }

    // Второй этап: отправка 2FA-кода
    twoFaResponse := send2FA(twoFaURL, twoFaCode)
    if twoFaResponse != nil {
        fmt.Println("✅ Авторизация завершена успешно")
        if token, ok := twoFaResponse["authentication"].(map[string]interface{})["token"]; ok {
            fmt.Printf("Токен: %s\n", token)
        }
    } else {
        fmt.Println("❌ Авторизация не удалась")
    }
}

File: practice/weaks/2-3/DAST/authirization/2fa_test.py
import requests
import pyotp

# Shared secret из QR-кода
SECRET = "LVNT2ARXHR3XUFLS"

# URL для логина и 2FA
AUTH_URL = "http://localhost:3000/rest/user/login"
TWO_FA_URL = "http://localhost:3000/rest/2fa/verify"

# Логин и пароль
USERNAME = "test@mail.com"
PASSWORD = "password01"

# Генерация 2FA-кода с использованием pyotp
def generate_2fa_code(secret):
    totp = pyotp.TOTP(secret)
    two_fa_code = totp.now()
    print(f"✅ Сгенерированный 2FA-код: {two_fa_code}")
    return two_fa_code

# Выполнение первого этапа авторизации (логин)
def login(auth_url, username, password):
    payload = {"email": username, "password": password}
    print(f"ℹ️  Отправляем запрос на логин с payload: {payload}")
    response = requests.post(auth_url, json=payload)
    print(f"ℹ️  Ответ от сервера на логин: {response.status_code}, {response.text}")
    if response.status_code == 200:
        print("✅ Успешный логин")
        return response.json()
    elif response.status_code == 401 and "totp_token_required" in response.text:
        print("ℹ️  Требуется 2FA-код")
        return response.json()  # Возвращаем ответ с tmpToken
    else:
        print(f"❌ Ошибка логина: {response.status_code}")
        print(response.text)
        return None

# Выполнение второго этапа авторизации (2FA)
def send_2fa(two_fa_url, two_fa_code, tmp_token):
    payload = {"totpToken": two_fa_code, "tmpToken": tmp_token}
    print(f"ℹ️  Отправляем запрос на 2FA с payload: {payload}")
    response = requests.post(two_fa_url, json=payload)
    print(f"ℹ️  Ответ от сервера на 2FA: {response.status_code}, {response.text}")
    if response.status_code == 200:
        print("✅ Успешная 2FA")
        return response.json()
    else:
        print(f"❌ Ошибка 2FA: {response.status_code}")
        print(response.text)
        return None

# Основной процесс авторизации
def main():
    # Первый этап: логин
    login_response = login(AUTH_URL, USERNAME, PASSWORD)
    if not login_response:
        return

    # Извлечение временного токена
    tmp_token = login_response.get("data", {}).get("tmpToken")
    if not tmp_token:
        print("❌ Не удалось получить временный токен")
        return
    print(f"ℹ️  Полученный временный токен: {tmp_token}")

    # Генерация 2FA-кода
    two_fa_code = generate_2fa_code(SECRET)

    # Второй этап: отправка 2FA-кода
    two_fa_response = send_2fa(TWO_FA_URL, two_fa_code, tmp_token)
    if two_fa_response:
        print("✅ Авторизация завершена успешно")
        print(f"Токен: {two_fa_response.get('authentication', {}).get('token')}")
    else:
        print("❌ Авторизация не удалась")

if __name__ == "__main__":
    main()
